src/base/Pool.hh

```diff
@@ -0,0 +1,219 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Pool.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "NumericLimits.hh"
+#include "Types.hh"
+
+// Proxy for defining specializations in a separate header that device-only
+// code can omit
+#define POOL_HOST_HEADER 1
+#if POOL_HOST_HEADER
+#    include <vector>
+#    include "DeviceVector.hh"
+#endif
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Data ownership flag
+enum class Ownership
+{
+    value,           //!< Ownership of the data, only on host
+    reference,       //!< Mutable reference to the data
+    const_reference, //!< Immutable reference to the data
+};
+
+//! Size/offset type for pool
+using pool_size_type = unsigned int;
+
+CELER_CONSTEXPR_FUNCTION size_type max_pool_size()
+{
+    return static_cast<size_type>(numeric_limits<pool_size_type>::max());
+}
+
+namespace detail
+{
+template<typename T, Ownership W>
+struct ownership_traits
+{
+    using SpanT          = Span<T>;
+    using reference_type = T&;
+};
+
+template<typename T>
+struct ownership_traits<T, Ownership::const_reference>
+{
+    using SpanT          = Span<const T>;
+    using reference_type = const T&;
+};
+} // namespace detail
+
+//---------------------------------------------------------------------------//
+template<class T>
+class PoolRange
```

Comment:
How is it different for `span`? (I.e. isn't it a `[Poll/Porta]Span` ?)

Reply:
Span can actually access the memory directly (pointer plus size); this is an offset that needs someone else's data pointer (the `Pool`'s) to get to data. It's just a pair of start/stop indices, which is the definition of a range.

Reply:
Actually I'm wrong, it's the definition of a slice (to use python nomenclature), not a container that mimics a sequence of numbers (which is really what a range is). So I should rename this to `PoolSlice` (or whatever substitute for `Pool`).

Reply:
Fair enough, this semantic also matches [std::slice](https://en.cppreference.com/w/cpp/numeric/valarray/slice)

Reply:
Ugh - slice only applies to the "never-to-used" `valarray` doesn't it?

Reply:
@sethrj I'm not sure how much I like `Portable` since that seems somewhat vague. I hate to mention this, but should we consider `View` or something more like that?

Reply:
> Ugh - slice only applies to the "never-to-used" valarray doesn't it?

Yep :) still it is the same concept :)

> should we consider `View`

`View` does not match well with the fact that the object sometimes owns and manage the data.

Reply:
That is true, but it does match the semantics of a `kokkos::View`.

Reply:
Yeh, I wouldn't get in too big a debate over it, but if it comes to it, I prefer `Pool` over `Portable` or `Porta`

Reply:
See https://celeritasproject.slack.com/archives/CSLM7G17V/p1612460624016800 for more debate :)

Reply:
Yeah @tmdelellis  we had a long discussion very early on about how to name `TrackView`. It's not congruent with Kokkos, but basically we have the following analogy:

 String  | Particle | Description
----- | ----- | -----
std::vector<char> | Pie<ParticleState, value> | Manages/owns data
std::span<char> | Pie<ParticleState, reference> | Low level access to the raw data
std::string_view | ParticleView | High level operation on the data

So here `View` has a similar analogy to the `std::view`, whereas Kokkos views are more like `std::span` + `std::vector`, aka this class.

Reply:
Wow @tmdelellis it's been 10 years since I looked at `valarray` and had totally forgotten that it has a `slice` object. But yes, aside from the `stride` which this doesn't have, the semantics are the same.